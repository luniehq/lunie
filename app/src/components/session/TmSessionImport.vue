<template>
  <SessionFrame :icon="`settings_backup_restore`">
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title">Recover with backup code</h2>
      <div class="session-main bottom-indent">
        <Steps :steps="[`Recover`, `Name`, `Password`]" active-step="Recover" />
        <TmFormGroup
          :error="$v.$error && $v.seed.$invalid"
          field-id="import-seed"
          field-label="Backup code"
        >
          <FieldSeed
            id="import-seed"
            :value="seed"
            :placeholder="
              isPolkadot
                ? 'Must be your seed phrase or private key hash'
                : 'Must be exactly 12 or 24 words'
            "
            @input="(val) => (seed = val)"
          />
          <TmFormMsg
            v-if="$v.seed.$error && !$v.seed.required"
            name="Seed"
            type="required"
          />
          <TmFormMsg
            v-else-if="$v.seed.$error && !$v.seed.seedHasCorrectLength"
            name="Seed"
            :type="isPolkadot ? 'incorrectPolkadotSeed' : 'words12or24'"
          />
          <TmFormMsg
            v-else-if="$v.seed.$error && !$v.seed.seedIsLowerCaseAndSpaces"
            name="Seed"
            :type="isPolkadot ? 'incorrectPolkadotSeed' : 'lowercaseAndSpaces'"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Next" type="submit" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { required } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmFormMsg from "common/TmFormMsg"
import FieldSeed from "common/TmFieldSeed"
import SessionFrame from "common/SessionFrame"
import { mapGetters } from "vuex"
import Steps from "../../ActionModal/components/Steps"
import { mnemonicValidate } from "@polkadot/util-crypto"

const has12or24words = (param) => {
  return (
    param && (param.split(` `).length === 12 || param.split(` `).length === 24)
  )
}
const isHex = (value) => {
  return /0x[0-9a-f]+/i.test(value)
}

const lowerCaseAndSpaces = (param) => {
  const seedWordsAreLowerCaseAndSpaces = /^([a-z]+\s)*[a-z]+$/g
  if (param.match(seedWordsAreLowerCaseAndSpaces)) {
    return param === param.match(seedWordsAreLowerCaseAndSpaces)[0]
  }
  return false
}
// exporting these for testing
export const isPolkadotHexSeed = (seed) => {
  return isHex(seed) && seed.length === 66
}

export const polkadotRawSeedValidate = (seed) => {
  return (seed.length > 0 && seed.length <= 32) || isPolkadotHexSeed(seed)
}

export const polkadotValidation = (seed) => {
  return mnemonicValidate(seed) || polkadotRawSeedValidate(seed)
}

export default {
  name: `session-import-backupcode`,
  components: {
    TmBtn,
    SessionFrame,
    FieldSeed,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps,
  },
  computed: {
    ...mapGetters([`recover`, `currentNetwork`]),
    seed: {
      get() {
        return this.$store.state.recover.seed
      },
      set(value) {
        this.$store.commit(`updateField`, {
          field: `seed`,
          value: value.trim(), // remove spaces from beginning and end of string
        })
      },
    },
    isPolkadot() {
      return this.currentNetwork.network_type === "polkadot"
    },
  },
  methods: {
    onSubmit() {
      this.$v.$touch()
      if (this.$v.seed.$invalid || this.$v.seed.$invalid) return
      this.$router.push("/recover/name")
    },
  },
  validations() {
    return {
      seed: {
        required,
        seedIsLowerCaseAndSpaces: (param) =>
          this.isPolkadot
            ? polkadotValidation(param)
            : lowerCaseAndSpaces(param),
        seedHasCorrectLength: (param) =>
          this.isPolkadot ? polkadotValidation(param) : has12or24words(param),
      },
    }
  },
}
</script>
