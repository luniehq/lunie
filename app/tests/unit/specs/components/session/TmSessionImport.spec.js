import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import TmSessionImport from "session/TmSessionImport"
import {
  isPolkadotHexSeed,
  polkadotRawSeedValidate,
  polkadotValidation,
} from "session/TmSessionImport"
jest.mock(`scripts/google-analytics.js`, () => () => {})
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})

describe(`TmSessionImport`, () => {
  let wrapper, $store, getters

  beforeEach(() => {
    getters = {
      connected: () => true,
      currentNetwork: {
        id: "kusama",
        network_type: "polkadot",
      },
    }
    $store = {
      state: {
        recover: {
          name: ``,
          seed: ``,
          password: ``,
          passwordConfirm: ``,
          warning: false,
        },
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => true),
      mutations: {
        updateField: jest.fn(),
      },
    }
    wrapper = shallowMount(TmSessionImport, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`validation should fail if seed is not filled in`, async () => {
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.seed.$error).toBe(true)
  })

  it(`validation should fail if seed is not 12 or 24 words long`, async () => {
    wrapper = shallowMount(TmSessionImport, {
      localVue,
      mocks: {
        $store: {
          state: {
            recover: { seed: `` },
          },
          getters: {
            network: "cosmos-hub-mainnet",
            currentNetwork: {
              id: "cosmos-hub-mainnet",
              network_type: "cosmos",
            },
          },
        },
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
    wrapper.vm.$store.state.recover.seed = `asdf asdf asdf asdf`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.seed.$error).toBe(true)
  })

  it(`should validate if seed is 12 or 24 words long`, async () => {
    wrapper = shallowMount(TmSessionImport, {
      localVue,
      mocks: {
        $store: {
          state: {
            recover: { seed: `` },
          },
          getters: {
            network: "cosmos-hub-mainnet",
            currentNetwork: {
              id: "cosmos-hub-mainnet",
              network_type: "cosmos",
            },
          },
        },
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
    wrapper.vm.$store.state.recover.seed = `asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.seed.$error).toBe(false)
    wrapper.vm.$store.state.recover.seed = `asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf`
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.seed.$error).toBe(false)
  })

  it(`should commit updateField on field change`, async () => {
    wrapper.setData({ seed: `asdf asdf asdf asdf` })
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `seed`,
      value: `asdf asdf asdf asdf`,
    })
  })

  it(`should go to /recover/name when submit the form`, async () => {
    wrapper = shallowMount(TmSessionImport, {
      localVue,
      mocks: {
        $store: {
          state: {
            recover: { seed: `` },
          },
          getters: {
            network: "cosmos-hub-mainnet",
            currentNetwork: {
              id: "cosmos-hub-mainnet",
              network_type: "cosmos",
            },
          },
        },
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
    wrapper.vm.$store.state.recover.seed = `asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf`
    wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith(`/recover/name`)
  })

  it(`should validate if a seed is a Kusama hex seed`, () => {
    let seed =
      "0x3ddaf335b6df385e5cfc6b78c9ce112afd8ac38f6794d29af3ee5f808dfae348"
    expect(isPolkadotHexSeed(seed)).toBe(true)
    seed = "ddaf335b6df385e5cfc6b78c9ce112afd8ac38f6794d29af3ee5f808dfae348123"
    expect(isPolkadotHexSeed(seed)).toBe(false)
  })

  it(`should validate if a seed is a Kusama raw seed`, () => {
    let seed = "0xd551233cd15fdb8b0dca468d38ac"
    expect(polkadotRawSeedValidate(seed)).toBe(true)
    seed = "d551233cd15fdb8b0dca468d38ac"
    expect(polkadotRawSeedValidate(seed)).toBe(true) // weird?? But it is on PolkadotUI ¯\_(ツ)_/¯
  })

  it(`should validate if a seed is a Kusama correct seed`, () => {
    let seed =
      "frame uphold talent kiwi chapter horn hen margin essay power frozen wisdom"
    expect(polkadotValidation(seed)).toBe(true)
    seed =
      "frame uphold talent kiwi chapter horn hen margin essay power frozen wisdo"
    expect(polkadotValidation(seed)).toBe(false)
  })
})
