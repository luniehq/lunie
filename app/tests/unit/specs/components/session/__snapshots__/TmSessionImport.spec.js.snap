// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionImport has the expected html structure 1`] = `
<sessionframe-stub
  icon="settings_backup_restore"
>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      Recover with backup code
    </h2>
     
    <div
      class="session-main bottom-indent"
    >
      <steps-stub
        activestep="Recover"
        steps="Recover,Name,Password"
      />
       
      <tmformgroup-stub
        fieldid="import-seed"
        fieldlabel="Backup code"
      >
        <fieldseed-stub
          id="import-seed"
          placeholder="Must be your seed phrase or private key hash"
          value=""
        />
         
        <!---->
      </tmformgroup-stub>
    </div>
     
    <div
      class="session-footer"
    >
      <tmbtn-stub
        customclass=""
        type="submit"
        value="Next"
      />
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
